File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int n = truck_weights.length;
        int[] bridge = new int[bridge_length];
        int[] visited = new int[truck_weights.length];
        int total = 0;
        for (int i = 0; i < truck_weights.length; i++) {
            total += truck_weights[i];
        }
        
        int x = 0;
        int cur = 0;
        int time = 0;
        int pass = 0;
        boolean left = true;
        for (int l = 0; l < truck_weights.length * bridge_length + 1; l++ ) {
            
            
            
            if (pass == total) {
                break;
            }
            if (cur == 0 && left && visited[x] == 0) {
                bridge[bridge_length - 1] = truck_weights[x];
                cur += truck_weights[x];
                visited[x] = 1;
                if (x < truck_weights.length - 1) {
                    x += 1;
                }
            } else {
                int tmp = bridge[0];
                if (tmp > 0) {
                    cur -= tmp;
                    pass += tmp;
                }
                if (cur + truck_weights[x] > weight) {
                    for (int i = 1; i < bridge_length; i++) {
                        bridge[i - 1] = bridge[i];
                    }
                    bridge[bridge_length - 1] = 0;
                } else {
                    for (int i = 1; i < bridge_length; i++) {
                        bridge[i - 1] = bridge[i];
                    }
                    bridge[bridge_length - 1] = 0;
                    if (left && visited[x] == 0) {
                        bridge[bridge_length - 1] = truck_weights[x];
                        cur += truck_weights[x];
                        visited[x] = 1;
                        if (x < truck_weights.length - 1) {
                            x += 1;
                        }
                    }
                }
            }
            time += 1;
            if (visited[visited.length - 1] == 1) {
                left = false;
            }
            
            // System.out.print(time + ": ");
            // for (int j = 0; j < bridge_length; j++) {
            //     System.out.print(bridge[j] + " ");
            // }
            // System.out.println();
        }
        
        answer = time; 
        return answer;
    }
}

File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    answer = 0
    n = len(truck_weights)
    bridge = [0] * bridge_length
    visited = [False] * n
    total = sum(truck_weights)
    
    index = 0
    cur = 0
    passed = 0
    left = True
    ## 2 10 [7, 7, 7, 7]과 같은 최악의 경우에 걸리는 시간은 (다리의 길이) X (총 트럭의 수) + 1이다.
    for i in range(bridge_length * n + 1):
        #이미 지나간 트럭의 무게와 총 트럭의 무게가 같으면
        if passed == total:
            break
        # 현재 다리 위에 아무것도 없고 남은 차량이 있다면
        if cur == 0 and left and visited[index] == False :
            bridge[bridge_length - 1] = truck_weights[index]
            cur += truck_weights[index]
            visited[index] = True
            # 마지막 차량보다 앞에 있는 순서의 차량일 경우
            if index < n - 1: 
                index += 1
        #다리 위에 트럭이 있으면
        else:
            tmp = bridge[0] #나가야할 차량을 내보내고
            if tmp > 0:
                cur -= tmp
                passed += tmp
            for i in range(1, bridge_length):
                bridge[i - 1] = bridge[i]
            bridge[bridge_length - 1] = 0
            # 다음에 들어올 차량의 무게와 현재 다리위에 있는 차량들의
            # 무게의 합이 다리가 견딜 수 있는 총 중량보다 작거나 같으면
            if cur + truck_weights[index] <= weight:
                #남은 차량이 있다면
                if left and visited[index] == False:
                    bridge[bridge_length - 1] = truck_weights[index]
                    cur += truck_weights[index]
                    visited[index] = True
                    if index < n - 1:
                        index += 1
                        
        #시간 재기
        answer += 1
        
        #마지막 차량이 다리위에 올라왔으면 남은 차량 없음
        if visited[n - 1]:
            left = False

    return answer